rest_framework/decorators.py

```diff
@@ -46,11 +47,19 @@ def decorator(func):
         allowed_methods = set(http_method_names) | {'options'}
         WrappedAPIView.http_method_names = [method.lower() for method in allowed_methods]
 
-        def handler(self, *args, **kwargs):
+        def sync_handler(self, *args, **kwargs):
             return func(*args, **kwargs)
 
+        async def async_handler(self, *args, **kwargs):
+            return await func(*args, **kwargs)
+
+        view_is_async = asyncio.iscoroutinefunction(func)
+
         for method in http_method_names:
-            setattr(WrappedAPIView, method.lower(), handler)
+            if view_is_async:
```

Comment:
Not important here, but it appears that Python doesn't optimize this

![image](https://user-images.githubusercontent.com/6127032/187811110-646cd1b6-a79d-4a4a-910d-eeac5b77e93c.png)

Reply:
I put the if statement outside the for loop. Thanks.